clients/drcachesim/tools/trace_filter.cpp

```diff
@@ -0,0 +1,217 @@
+}
+
+std::unique_ptr<std::ostream>
+trace_filter_t::get_writer(const std::string &path, int verbosity)
+{
+    std::unique_ptr<std::ostream> ofile;
+#ifdef HAS_ZLIB
+    if (input_file_format == FILE_FORMAT_GZIP) {
+        VPRINT(this, 3, "Using the gzip writer for %s\n", path.c_str());
+        return std::unique_ptr<std::ostream>(new gzip_ostream_t(path));
+    }
+#endif
+    assert(input_file_format == FILE_FORMAT_UNKNOWN);
+    VPRINT(this, 3, "Using the default writer for %s\n", path.c_str());
+    return std::unique_ptr<std::ostream>(new std::ofstream(path, std::ofstream::binary));
+}
+
+bool
+trace_filter_t::init_file_reader_writer(const std::string &trace_dir,
+                                        const std::string &output_dir, int verbosity)
+{
+    if (trace_dir.empty() || output_dir.empty()) {
+        ERRMSG("Trace dir or output dir name is empty\n");
+        return false;
+    }
+    if (!directory_iterator_t::is_directory(trace_dir) ||
+        !directory_iterator_t::is_directory(output_dir)) {
+        ERRMSG("Trace dir or output dir is not a directory\n");
+        return false;
+    }
+    directory_iterator_t end;
+    directory_iterator_t iter(trace_dir);
+    if (!iter) {
+        ERRMSG("Failed to list directory %s: %s", trace_dir.c_str(),
+               iter.error_string().c_str());
+        return false;
+    }
+    for (; iter != end; ++iter) {
+        const std::string fname = *iter;
+        if (fname == "." || fname == "..")
+            continue;
+        const std::string trace_path = trace_dir + DIRSEP + fname;
+        const std::string output_path = output_dir + DIRSEP + fname;
+        std::unique_ptr<trace_entry_reader_t> reader = get_reader(trace_path, verbosity);
+        if (!reader) {
+            error_string_ = "Could not get a reader for " + trace_path;
+            return false;
+        }
+
+        std::unique_ptr<std::ostream> writer = get_writer(output_path, verbosity);
+        if (!writer) {
+            error_string_ = "Could not get a writer for " + output_path;
+            return false;
+        }
+
+        thread_data_.push_back(shard_data_t(static_cast<int>(thread_data_.size()),
+                                            std::move(reader), std::move(writer),
+                                            trace_path, output_path));
+    }
+    // Like raw2trace, we use a simple round-robin static work assigment.  This
+    // could be improved later with dynamic work queue for better load balancing.
+    if (worker_count_ <= 0)
+        worker_count_ = std::thread::hardware_concurrency();
+    worker_tasks_.resize(worker_count_);
+    int worker = 0;
+    for (size_t i = 0; i < thread_data_.size(); ++i) {
+        VPRINT(this, 2, "Worker %d assigned trace shard %zd\n", worker, i);
+        worker_tasks_[worker].push_back(&thread_data_[i]);
+        thread_data_[i].worker = worker;
+        worker = (worker + 1) % worker_count_;
+    }
```

Comment:
This is all very similar to analyzer_t::init_file_reader, identical in many places.  Maybe this should be a sub-class with just get_reader() overridden to also create an ostream.

Reply:
See concerns above about sub-classing.

We could put it in a common library, but then `shard_data_t` is different here: has output file and stream objects too.

Reply:
Converted analyzer_t to a template in https://github.com/DynamoRIO/dynamorio/pull/5720.